lib/trino-hive-formats/src/main/java/io/trino/hive/formats/line/json/JsonDeserializer.java

```diff
@@ -0,0 +1,838 @@
+                decimalType.writeLong(builder, bigDecimal.unscaledValue().longValueExact());
+            }
+            else {
+                decimalType.writeObject(builder, Int128.valueOf(bigDecimal.unscaledValue()));
+            }
+        }
+    }
+
+    private static class RealDecoder
+            extends Decoder
+    {
+        public RealDecoder()
+        {
+            super(REAL);
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            REAL.writeLong(builder, floatToRawIntBits(parser.getFloatValue()));
+        }
+    }
+
+    private static class DoubleDecoder
+            extends Decoder
+    {
+        public DoubleDecoder()
+        {
+            super(DOUBLE);
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            DOUBLE.writeDouble(builder, parser.getDoubleValue());
+        }
+    }
+
+    private static class DateDecoder
+            extends Decoder
+    {
+        public DateDecoder()
+        {
+            super(DATE);
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            DATE.writeLong(builder, toIntExact(parseHiveDate(parser.getText()).toEpochDay()));
+        }
+    }
+
+    private static class TimestampDecoder
+            extends Decoder
+    {
+        private final TimestampType timestampType;
+        private final Function<String, DecodedTimestamp> timestampParser;
+
+        public TimestampDecoder(TimestampType timestampType, Function<String, DecodedTimestamp> timestampParser)
+        {
+            super(timestampType);
+            this.timestampType = timestampType;
+            this.timestampParser = timestampParser;
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            DecodedTimestamp timestamp = timestampParser.apply(parser.getText());
+            createTimestampEncoder(timestampType, UTC).write(timestamp, builder);
+        }
+    }
+
+    private static class VarbinaryDecoder
+            extends Decoder
+    {
+        // This must be created per instance because it is not thread safe
+        private final CharsetDecoder charsetDecoder = createCharsetDecoder();
+
+        public VarbinaryDecoder()
+        {
+            super(VARBINARY);
+        }
+
+        @Override
+        void decodeValue(LineBuffer lineBuffer, JsonParser parser, BlockBuilder builder)
+                throws IOException
+        {
+            VARBINARY.writeSlice(builder, parseBinary(parser.getText(), charsetDecoder));
+        }
+
+        private static Slice parseBinary(String value, CharsetDecoder charsetDecoder1)
```

Comment:
Rename `charsetDecoder1` or probably make this method non-static and use the field